File: server/crypto/master_vault_ledger.py
"""
Master Vault & 51/49 Cap Cryptographic Ledger Engine
Token Identifier: 9898048483
Total Supply: 989,804,848,300

Architecture:
- 100% initial supply allocated to Admin Master Vault (989,804,848,300 tokens).
- 51% strictly locked Admin Reserve (504,800,472,633 tokens) untouchable by public minting.
- 49% strict Public Distribution Cap (485,004,375,667 tokens max).
- Device Onboarding: 1,000 tokens credited per unique verified device upon registration.
- Auto-pause trigger when the 49% public cap is reached, with Admin multi-sig / cryptographic override.
- Immutable hash-chained audit ledger with SHA-256 state commitments.
"""

import time
import hashlib
import json
import threading
import logging
from typing import Dict, Any, Optional, Tuple, List
from dataclasses import dataclass, field, asdict

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("MasterVaultLedger")


# ---------------------------------------------------------------------------
# Constants & Token Economics Matrix
# ---------------------------------------------------------------------------
TOKEN_ID: str = "9898048483"
TOTAL_SUPPLY: int = 989_804_848_300  # 989.8B tokens

# 51% Locked Admin Reserve (Cannot be distributed to public via standard registration/rewards)
ADMIN_RESERVE_PERCENT: float = 0.51
LOCKED_ADMIN_RESERVE: int = 504_800_472_633

# 49% Public Distribution Cap
PUBLIC_CAP_PERCENT: float = 0.49
MAX_PUBLIC_DISTRIBUTION: int = 485_004_375_667

# Per Device Registration Incentive
DEVICE_REGISTRATION_REWARD: int = 1_000

ADMIN_MASTER_VAULT_ADDRESS: str = "vault_master_9898048483_admin_enclave"


@dataclass
class LedgerTransaction:
    tx_id: str
    from_address: str
    to_address: str
    amount: int
    tx_type: str  # "DEVICE_REGISTRATION", "TRANSFER", "ADMIN_ALLOCATION", "STAKING_REWARD"
    device_id: Optional[str]
    timestamp: float
    prev_hash: str
    tx_hash: str
    metadata: Dict[str, Any] = field(default_factory=dict)


@dataclass
class DeviceRecord:
    device_id: str
    wallet_address: str
    pqc_pubkey_hash: str
    registered_at: float
    initial_grant: int
    attestation_verified: bool


class MasterVaultLedgerEngine:
    """
    Cryptographically secure, thread-safe Ledger Engine governing token 9898048483
    with strict 51/49 reserve enforcement and automated device registration.
    """

    def __init__(
        self,
        admin_pubkey_hash: str = "a9898048483_secp256k1_pqc_master_key",
        storage_filepath: Optional[str] = "logs/master_vault_ledger.json",
    ) -> None:
        self.lock = threading.RLock()
        self.admin_pubkey_hash = admin_pubkey_hash
        self.storage_filepath = storage_filepath

        # Vault & Balance State
        self.admin_vault_balance: int = TOTAL_SUPPLY
        self.total_public_distributed: int = 0
        self.wallets: Dict[str, int] = {
            ADMIN_MASTER_VAULT_ADDRESS: TOTAL_SUPPLY
        }

        # Registered Device Tracking (device_id -> DeviceRecord)
        self.registered_devices: Dict[str, DeviceRecord] = {}
        # Device Hardware / Address Deduplication
        self.device_wallet_map: Dict[str, str] = {}

        # Issuance Controls
        self.is_issuance_paused: bool = False
        self.admin_manual_override: bool = False

        # Hash Chained Ledger
        self.transactions: List[LedgerTransaction] = []
        self.last_block_hash: str = "0" * 64

        # Initialize Genesis Transaction
        self._initialize_genesis()

    def _initialize_genesis(self) -> None:
        """Records Genesis allocation of 100% total supply into the Admin Master Vault."""
        with self.lock:
            genesis_payload = {
                "token_id": TOKEN_ID,
                "total_supply": TOTAL_SUPPLY,
                "admin_reserve": LOCKED_ADMIN_RESERVE,
                "public_cap": MAX_PUBLIC_DISTRIBUTION,
                "timestamp": time.time(),
            }
            payload_str = json.dumps(genesis_payload, sort_keys=True)
            genesis_hash = hashlib.sha256(f"GENESIS|{payload_str}".encode("utf-8")).hexdigest()

            genesis_tx = LedgerTransaction(
                tx_id="tx_genesis_9898048483",
                from_address="0x0000000000000000000000000000000000000000",
                to_address=ADMIN_MASTER_VAULT_ADDRESS,
                amount=TOTAL_SUPPLY,
                tx_type="GENESIS_MINT",
                device_id=None,
                timestamp=time.time(),
                prev_hash=self.last_block_hash,
                tx_hash=genesis_hash,
                metadata=genesis_payload,
            )
            self.transactions.append(genesis_tx)
            self.last_block_hash = genesis_hash
            logger.info(
                f"[Genesis Initialized] 100% Supply ({TOTAL_SUPPLY:,} Tokens) minted to Master Vault {ADMIN_MASTER_VAULT_ADDRESS}"
            )

    def _compute_tx_hash(
        self, prev_hash: str, from_addr: str, to_addr: str, amount: int, timestamp: float, device_id: Optional[str]
    ) -> str:
        """Generates a SHA-256 cryptographic link in the ledger chain."""
        raw = f"{prev_hash}|{from_addr}|{to_addr}|{amount}|{timestamp}|{device_id or ''}".encode("utf-8")
        return hashlib.sha256(raw).hexdigest()

    def register_device(
        self,
        device_id: str,
        wallet_address: str,
        pqc_pubkey_hash: str,
        attestation_data: Optional[Dict[str, Any]] = None,
    ) -> Tuple[bool, str, Optional[Dict[str, Any]]]:
        """
        Registers a new valid device and deducts exactly 1,000 tokens from the Admin Master Vault
        to transfer to the user's wallet address, respecting the 49% public distribution cap.
        """
        with self.lock:
            # 1. Deduplication Checks (Prevent Sybil / double registration)
            if device_id in self.registered_devices:
                return False, f"Device '{device_id}' is already registered and claimed initial grant.", None

            if wallet_address in self.device_wallet_map:
                prev_dev = self.device_wallet_map[wallet_address]
                return False, f"Wallet '{wallet_address}' is already linked to device '{prev_dev}'.", None

            # 2. Check 49% Public Distribution Cap & Pause Status
            next_distribution_total = self.total_public_distributed + DEVICE_REGISTRATION_REWARD
            
            # Check if cap is reached
            if next_distribution_total > MAX_PUBLIC_DISTRIBUTION:
                self.is_issuance_paused = True
                if not self.admin_manual_override:
                    logger.warning(
                        f"[Public Cap Reached] Public distribution ceiling of {MAX_PUBLIC_DISTRIBUTION:,} reached. Auto-pausing."
                    )
                    return (
                        False,
                        f"CAP_REACHED: 49% Public Distribution Cap ({MAX_PUBLIC_DISTRIBUTION:,} tokens) exceeded or reached. Public token issuance is paused to protect the 51% Locked Admin Reserve.",
                        None,
                    )

            # Check if paused without manual admin override
            if self.is_issuance_paused and not self.admin_manual_override:
                return False, f"CAP_REACHED: Public token issuance is currently paused because 49% distribution cap was reached or exceeded.", None

            # 3. Ensure Master Vault maintains the 51% Locked Admin Reserve
            remaining_vault = self.admin_vault_balance - DEVICE_REGISTRATION_REWARD
            if remaining_vault < LOCKED_ADMIN_RESERVE and not self.admin_manual_override:
                return False, f"Transaction violates 51% Locked Admin Reserve ({LOCKED_ADMIN_RESERVE:,} tokens).", None

            # 4. Atomic Balance Updates
            self.admin_vault_balance -= DEVICE_REGISTRATION_REWARD
            self.total_public_distributed += DEVICE_REGISTRATION_REWARD
            self.wallets[ADMIN_MASTER_VAULT_ADDRESS] = self.admin_vault_balance
            self.wallets[wallet_address] = self.wallets.get(wallet_address, 0) + DEVICE_REGISTRATION_REWARD

            # 5. Record Device
            timestamp = time.time()
            device_record = DeviceRecord(
                device_id=device_id,
                wallet_address=wallet_address,
                pqc_pubkey_hash=pqc_pubkey_hash,
                registered_at=timestamp,
                initial_grant=DEVICE_REGISTRATION_REWARD,
                attestation_verified=True if attestation_data else False,
            )
            self.registered_devices[device_id] = device_record
            self.device_wallet_map[wallet_address] = device_id

            # 6. Append Immutable Transaction to Hash Chain
            tx_id = f"tx_dev_reg_{int(timestamp)}_{len(self.transactions)}"
            tx_hash = self._compute_tx_hash(
                self.last_block_hash,
                ADMIN_MASTER_VAULT_ADDRESS,
                wallet_address,
                DEVICE_REGISTRATION_REWARD,
                timestamp,
                device_id,
            )

            tx = LedgerTransaction(
                tx_id=tx_id,
                from_address=ADMIN_MASTER_VAULT_ADDRESS,
                to_address=wallet_address,
                amount=DEVICE_REGISTRATION_REWARD,
                tx_type="DEVICE_REGISTRATION",
                device_id=device_id,
                timestamp=timestamp,
                prev_hash=self.last_block_hash,
                tx_hash=tx_hash,
                metadata={
                    "grant_amount": DEVICE_REGISTRATION_REWARD,
                    "pqc_pubkey_hash": pqc_pubkey_hash,
                    "public_distributed_so_far": self.total_public_distributed,
                    "remaining_vault_balance": self.admin_vault_balance,
                },
            )
            self.transactions.append(tx)
            self.last_block_hash = tx_hash

            # 7. Check if 49% cap is now reached after this transaction
            if self.total_public_distributed >= MAX_PUBLIC_DISTRIBUTION:
                self.is_issuance_paused = True
                logger.info("[Cap Trigger] 49% Distribution Cap threshold met. Auto-pausing standard issuance.")

            logger.info(
                f"[Device Registered] 1,000 tokens transferred to {wallet_address} for device {device_id}. TxHash: {tx_hash[:12]}..."
            )

            return True, "Device successfully registered. 1,000 tokens credited from Admin Master Vault.", {
                "tx_id": tx_id,
                "tx_hash": tx_hash,
                "device_id": device_id,
                "wallet_address": wallet_address,
                "credited_amount": DEVICE_REGISTRATION_REWARD,
                "wallet_balance": self.wallets[wallet_address],
                "admin_vault_remaining": self.admin_vault_balance,
                "total_public_distributed": self.total_public_distributed,
                "public_cap_remaining": max(0, MAX_PUBLIC_DISTRIBUTION - self.total_public_distributed),
                "is_issuance_paused": self.is_issuance_paused,
            }

    def transfer(
        self,
        from_address: str,
        to_address: str,
        amount: int,
        signature_proof: Optional[str] = None,
    ) -> Tuple[bool, str, Optional[Dict[str, Any]]]:
        """
        Executes standard token transfers between accounts while respecting reserve constraints.
        """
        with self.lock:
            if amount <= 0:
                return False, "Transfer amount must be strictly positive.", None

            from_bal = self.wallets.get(from_address, 0)
            if from_bal < amount:
                return False, f"Insufficient funds. Balance: {from_bal:,}, Required: {amount:,}.", None

            # If sender is Admin Master Vault, ensure 51% reserve is not violated
            if from_address == ADMIN_MASTER_VAULT_ADDRESS:
                if (from_bal - amount) < LOCKED_ADMIN_RESERVE and not self.admin_manual_override:
                    return False, f"Transfer rejected: Violates 51% Admin Reserve ({LOCKED_ADMIN_RESERVE:,} tokens).", None

            # Execute Transfer
            self.wallets[from_address] = from_bal - amount
            self.wallets[to_address] = self.wallets.get(to_address, 0) + amount

            if from_address == ADMIN_MASTER_VAULT_ADDRESS:
                self.admin_vault_balance = self.wallets[ADMIN_MASTER_VAULT_ADDRESS]
                self.total_public_distributed += amount

            timestamp = time.time()
            tx_id = f"tx_tf_{int(timestamp)}_{len(self.transactions)}"
            tx_hash = self._compute_tx_hash(
                self.last_block_hash, from_address, to_address, amount, timestamp, None
            )

            tx = LedgerTransaction(
                tx_id=tx_id,
                from_address=from_address,
                to_address=to_address,
                amount=amount,
                tx_type="TRANSFER",
                device_id=None,
                timestamp=timestamp,
                prev_hash=self.last_block_hash,
                tx_hash=tx_hash,
                metadata={"signature_proof": signature_proof or "VALIDATED_PQC_SESSION"},
            )
            self.transactions.append(tx)
            self.last_block_hash = tx_hash

            return True, "Transfer successfully settled on-chain.", {
                "tx_id": tx_id,
                "tx_hash": tx_hash,
                "from_address": from_address,
                "to_address": to_address,
                "amount": amount,
                "sender_balance": self.wallets[from_address],
                "receiver_balance": self.wallets[to_address],
            }

    def transfer_tokens(
        self,
        from_address: str,
        to_address: str,
        amount: float,
        signature_proof: Optional[str] = None,
    ) -> Tuple[bool, str, Optional[Dict[str, Any]]]:
        return self.transfer(from_address, to_address, int(amount), signature_proof)

    def set_admin_manual_override(
        self, admin_signature: str, override_enabled: bool, unpause: bool = True
    ) -> Tuple[bool, str]:
        """
        Allows authorized Admin to manually toggle issuance pause state and emergency overrides.
        """
        with self.lock:
            # Simple cryptographic/signature placeholder verification
            if not admin_signature or len(admin_signature) < 16:
                return False, "Invalid Admin signature authorization."

            self.admin_manual_override = override_enabled
            if unpause:
                self.is_issuance_paused = False

            action = "ENABLED" if override_enabled else "DISABLED"
            logger.info(f"[Admin Override] Manual override set to {action}. Issuance Paused: {self.is_issuance_paused}")
            return True, f"Admin override {action}. Issuance active: {not self.is_issuance_paused}"

    def get_ledger_state(self) -> Dict[str, Any]:
        """
        Returns full real-time metrics of the 51/49 token economy.
        """
        with self.lock:
            distributed_pct = (self.total_public_distributed / TOTAL_SUPPLY) * 100.0
            vault_pct = (self.admin_vault_balance / TOTAL_SUPPLY) * 100.0

            return {
                "token_id": TOKEN_ID,
                "total_supply": TOTAL_SUPPLY,
                "admin_master_vault_address": ADMIN_MASTER_VAULT_ADDRESS,
                "admin_master_vault_balance": self.admin_vault_balance,
                "admin_vault_percentage": f"{vault_pct:.4f}%",
                "locked_admin_reserve": LOCKED_ADMIN_RESERVE,
                "locked_admin_reserve_target": "51.0000%",
                "max_public_distribution_cap": MAX_PUBLIC_DISTRIBUTION,
                "public_distribution_cap_target": "49.0000%",
                "total_public_distributed": self.total_public_distributed,
                "public_distributed_tokens": float(self.total_public_distributed),
                "public_distributed_percentage": f"{distributed_pct:.4f}%",
                "remaining_public_allowance": max(0, MAX_PUBLIC_DISTRIBUTION - self.total_public_distributed),
                "current_vault_balance": float(self.admin_vault_balance),
                "total_registered_devices": len(self.registered_devices),
                "device_registration_grant": DEVICE_REGISTRATION_REWARD,
                "is_issuance_paused": self.is_issuance_paused,
                "admin_manual_override": self.admin_manual_override,
                "total_ledger_transactions": len(self.transactions),
                "last_block_hash": self.last_block_hash,
                "status": "OPERATIONAL" if not self.is_issuance_paused else "PAUSED_CAP_REACHED",
            }

    def register_device_and_grant(
        self,
        hwid_hash: str,
        wallet_address: Optional[str] = None,
        user_wallet_address: Optional[str] = None,
        device_model: str = "Android Device",
        pqc_pubkey_hash: Optional[str] = None,
        attestation_data: Optional[Dict[str, Any]] = None,
    ) -> Tuple[bool, str, Optional[Dict[str, Any]]]:
        addr = wallet_address or user_wallet_address or f"0x{hashlib.sha256(hwid_hash.encode()).hexdigest()}"
        pubkey_hash = pqc_pubkey_hash or hashlib.sha256(f"PUBKEY_{addr}".encode()).hexdigest()
        att = attestation_data or {"device_model": device_model}
        return self.register_device(
            device_id=hwid_hash,
            wallet_address=addr,
            pqc_pubkey_hash=pubkey_hash,
            attestation_data=att,
        )

    def get_vault_status(self) -> Dict[str, Any]:
        """Returns full real-time metrics of the 51/49 token economy (alias for get_ledger_state)."""
        return self.get_ledger_state()

    def get_balance(self, wallet_address: str) -> int:
        """Queries the balance of any wallet address."""
        with self.lock:
            return self.wallets.get(wallet_address, 0)

    def verify_ledger_integrity(self) -> Tuple[bool, str]:
        """
        Audits the entire transaction chain from Genesis to current block to prove zero tampering.
        """
        with self.lock:
            prev_hash = "0" * 64
            for idx, tx in enumerate(self.transactions):
                if tx.prev_hash != prev_hash:
                    return False, f"Hash chain broken at index {idx} (Tx {tx.tx_id})."
                
                # Check Genesis
                if idx == 0:
                    prev_hash = tx.tx_hash
                    continue

                recalculated = self._compute_tx_hash(
                    tx.prev_hash,
                    tx.from_address,
                    tx.to_address,
                    tx.amount,
                    tx.timestamp,
                    tx.device_id,
                )
                if recalculated != tx.tx_hash:
                    return False, f"Cryptographic integrity mismatch at index {idx} (Tx {tx.tx_id})."

                prev_hash = tx.tx_hash

            return True, f"All {len(self.transactions)} ledger transactions verified with SHA-256 integrity."


# Global Singleton Instance
master_vault_ledger = MasterVaultLedgerEngine()

File: server/crypto/master_vault_ledger.ts
import crypto from 'crypto';
import { auditLogger } from '../services/token_audit_logger.js';

export const TOKEN_ID = '9898048483';
export const TOTAL_SUPPLY = 989_804_848_300; // 989.8B tokens
export const LOCKED_ADMIN_RESERVE = 504_800_472_633; // 51%
export const MAX_PUBLIC_DISTRIBUTION = 485_004_375_667; // 49%
export const DEVICE_REGISTRATION_REWARD = 1_000;
export const ADMIN_MASTER_VAULT_ADDRESS = 'vault_master_9898048483_admin_enclave';

export interface LedgerTx {
  txId: string;
  fromAddress: string;
  toAddress: string;
  amount: number;
  txType: string;
  deviceId?: string;
  timestamp: number;
  prevHash: string;
  txHash: string;
  metadata?: Record<string, any>;
}

export interface DeviceRecord {
  deviceId: string;
  walletAddress: string;
  pqcPubkeyHash: string;
  registeredAt: number;
  initialGrant: number;
}

class MasterVaultLedgerEngine {
  private adminVaultBalance: number = TOTAL_SUPPLY;
  private totalPublicDistributed: number = 0;
  private wallets: Map<string, number> = new Map();
  private registeredDevices: Map<string, DeviceRecord> = new Map();
  private deviceWalletMap: Map<string, string> = new Map();
  private isIssuancePaused: boolean = false;
  private adminManualOverride: boolean = false;
  private transactions: LedgerTx[] = [];
  private lastBlockHash: string = '0'.repeat(64);

  constructor() {
    this.wallets.set(ADMIN_MASTER_VAULT_ADDRESS, TOTAL_SUPPLY);
    this.initializeGenesis();
  }

  private initializeGenesis() {
    const timestamp = Date.now();
    const payload = {
      tokenId: TOKEN_ID,
      totalSupply: TOTAL_SUPPLY,
      lockedReserve: LOCKED_ADMIN_RESERVE,
      publicCap: MAX_PUBLIC_DISTRIBUTION,
    };
    const genesisHash = crypto
      .createHash('sha256')
      .update(`GENESIS|${JSON.stringify(payload)}`)
      .digest('hex');

    const genesisTx: LedgerTx = {
      txId: 'tx_genesis_9898048483',
      fromAddress: '0x0000000000000000000000000000000000000000',
      toAddress: ADMIN_MASTER_VAULT_ADDRESS,
      amount: TOTAL_SUPPLY,
      txType: 'GENESIS_MINT',
      timestamp,
      prevHash: this.lastBlockHash,
      txHash: genesisHash,
      metadata: payload,
    };

    this.transactions.push(genesisTx);
    this.lastBlockHash = genesisHash;
  }

  private computeTxHash(
    prevHash: string,
    from: string,
    to: string,
    amount: number,
    timestamp: number,
    deviceId?: string
  ): string {
    return crypto
      .createHash('sha256')
      .update(`${prevHash}|${from}|${to}|${amount}|${timestamp}|${deviceId || ''}`)
      .digest('hex');
  }

  public registerDevice(
    deviceId: string,
    walletAddress: string,
    pqcPubkeyHash: string
  ): { success: boolean; message: string; data?: any } {
    // 1. Deduplication
    if (this.registeredDevices.has(deviceId)) {
      return { success: false, message: `Device ${deviceId} is already registered.` };
    }
    if (this.deviceWalletMap.has(walletAddress)) {
      return { success: false, message: `Wallet ${walletAddress} is already registered to a device.` };
    }

    // 2. Cap Check
    const nextTotal = this.totalPublicDistributed + DEVICE_REGISTRATION_REWARD;
    if (nextTotal > MAX_PUBLIC_DISTRIBUTION) {
      this.isIssuancePaused = true;
      if (!this.adminManualOverride) {
        return {
          success: false,
          message: `Registration paused: 49% Public Distribution Cap (${MAX_PUBLIC_DISTRIBUTION.toLocaleString()} tokens) reached.`,
        };
      }
    }

    if (this.isIssuancePaused && !this.adminManualOverride) {
      return { success: false, message: 'Public token issuance is currently paused.' };
    }

    // 3. 51% Locked Admin Reserve Safeguard
    const remainingVault = this.adminVaultBalance - DEVICE_REGISTRATION_REWARD;
    if (remainingVault < LOCKED_ADMIN_RESERVE && !this.adminManualOverride) {
      return {
        success: false,
        message: `Transaction violates 51% Locked Admin Reserve (${LOCKED_ADMIN_RESERVE.toLocaleString()} tokens).`,
      };
    }

    // 4. Update balances
    this.adminVaultBalance -= DEVICE_REGISTRATION_REWARD;
    this.totalPublicDistributed += DEVICE_REGISTRATION_REWARD;
    this.wallets.set(ADMIN_MASTER_VAULT_ADDRESS, this.adminVaultBalance);

    const currentWalletBal = this.wallets.get(walletAddress) || 0;
    this.wallets.set(walletAddress, currentWalletBal + DEVICE_REGISTRATION_REWARD);

    // 5. Store record
    const timestamp = Date.now();
    const deviceRecord: DeviceRecord = {
      deviceId,
      walletAddress,
      pqcPubkeyHash,
      registeredAt: timestamp,
      initialGrant: DEVICE_REGISTRATION_REWARD,
    };
    this.registeredDevices.set(deviceId, deviceRecord);
    this.deviceWalletMap.set(walletAddress, deviceId);

    // 6. Chain Ledger Transaction
    const txId = `tx_reg_${timestamp}_${this.transactions.length}`;
    const txHash = this.computeTxHash(
      this.lastBlockHash,
      ADMIN_MASTER_VAULT_ADDRESS,
      walletAddress,
      DEVICE_REGISTRATION_REWARD,
      timestamp,
      deviceId
    );

    const tx: LedgerTx = {
      txId,
      fromAddress: ADMIN_MASTER_VAULT_ADDRESS,
      toAddress: walletAddress,
      amount: DEVICE_REGISTRATION_REWARD,
      txType: 'DEVICE_REGISTRATION',
      deviceId,
      timestamp,
      prevHash: this.lastBlockHash,
      txHash,
      metadata: {
        pqcPubkeyHash,
        totalPublicDistributed: this.totalPublicDistributed,
      },
    };

    this.transactions.push(tx);
    this.lastBlockHash = txHash;

    // Log to encrypted audit logger
    auditLogger.recordEvent('TOKEN_TRANSFER', ADMIN_MASTER_VAULT_ADDRESS, {
      type: 'DEVICE_REGISTRATION_GRANT',
      to: walletAddress,
      amount: DEVICE_REGISTRATION_REWARD,
      deviceId,
      txHash,
    });

    if (this.totalPublicDistributed >= MAX_PUBLIC_DISTRIBUTION) {
      this.isIssuancePaused = true;
    }

    return {
      success: true,
      message: 'Device successfully registered. 1,000 tokens credited from Admin Master Vault.',
      data: {
        txId,
        txHash,
        deviceId,
        walletAddress,
        creditedAmount: DEVICE_REGISTRATION_REWARD,
        walletBalance: this.wallets.get(walletAddress),
        adminVaultRemaining: this.adminVaultBalance,
        totalPublicDistributed: this.totalPublicDistributed,
        publicCapRemaining: Math.max(0, MAX_PUBLIC_DISTRIBUTION - this.totalPublicDistributed),
        isIssuancePaused: this.isIssuancePaused,
      },
    };
  }

  public getLedgerMetrics() {
    const distributedPct = (this.totalPublicDistributed / TOTAL_SUPPLY) * 100;
    const vaultPct = (this.adminVaultBalance / TOTAL_SUPPLY) * 100;

    return {
      tokenId: TOKEN_ID,
      totalSupply: TOTAL_SUPPLY,
      adminMasterVaultAddress: ADMIN_MASTER_VAULT_ADDRESS,
      adminMasterVaultBalance: this.adminVaultBalance,
      adminVaultPercentage: `${vaultPct.toFixed(4)}%`,
      lockedAdminReserve: LOCKED_ADMIN_RESERVE,
      lockedAdminReserveTarget: '51.0000%',
      maxPublicDistributionCap: MAX_PUBLIC_DISTRIBUTION,
      publicDistributionCapTarget: '49.0000%',
      totalPublicDistributed: this.totalPublicDistributed,
      publicDistributedPercentage: `${distributedPct.toFixed(4)}%`,
      remainingPublicAllowance: Math.max(0, MAX_PUBLIC_DISTRIBUTION - this.totalPublicDistributed),
      totalRegisteredDevices: this.registeredDevices.size,
      deviceRegistrationGrant: DEVICE_REGISTRATION_REWARD,
      isIssuancePaused: this.isIssuancePaused,
      adminManualOverride: this.adminManualOverride,
      totalLedgerTransactions: this.transactions.length,
      lastBlockHash: this.lastBlockHash,
      status: !this.isIssuancePaused ? 'OPERATIONAL' : 'PAUSED_CAP_REACHED',
    };
  }

  public setAdminOverride(enabled: boolean, unpause: boolean = true) {
    this.adminManualOverride = enabled;
    if (unpause) {
      this.isIssuancePaused = false;
    }
    return {
      success: true,
      adminManualOverride: this.adminManualOverride,
      isIssuancePaused: this.isIssuancePaused,
    };
  }

  public getBalance(walletAddress: string): number {
    return this.wallets.get(walletAddress) || 0;
  }
}

export const masterVaultLedger = new MasterVaultLedgerEngine();
